# PAMmaker: pipelines creating allelic presence-absence matrices from results of gene screen

This repository consists of Python and R scripts that produce an allelic presence-absence matrix (PAM) from the outcome of a gene screen on bacterial samples (strains/isolates). Assuming _m_ alleles are identified in _n_ samples, the allelic PAM _**A** = (a<sub>ij</sub>)_ is an n-by-m binary matrix, where _a<sub>ij</sub> = 1_ when the _j_-th allele is present in the _i_-th sample, and _a<sub>ij</sub> = 0_ otherwise. This tool assumes the gene screen process produces files in the output format of [SRST2](github.com/katholt/srst2) and [screen\_genes\_in\_assemblies](github.com/wanyuac/screen_genes_in_assemblies).

## Installation
```
git clone https://github.com/wanyuac/PAMmaker.git
```

### Dependencies

* [R](https://www.r-project.org) (>=3.0)
* Python versions 2 and 3 compatible
* Linux bash

### Sub-directories

* reliability\_assessment: scripts used for evaluating reliability of allele calls
* unicity\_assessment: scripts used for collecting evidence for identifying co-occurrence of alleles of the same gene in each sample

## Usage

This section demonstrates a typical procedure for processing SRST2-compatible results.

### 1. Gene screen
Assuming a user uses SRST2 for screening antimicrobial resistance genes (ARGs) in bacterial samples, then an example SRST2 command line is (see SRST2 manual for details):

```
python srst2/scripts/slurm_srst2.py --script srst2/scripts/srst2.py --output demo --input_pe Reads/*_[1,2].fastq.gz --walltime '0-8:0:0' --threads 4 --memory 8192 --rundir ARGs --other_args "--gene_db srst2/data/ARGannot_r2.fasta --save_scores --report_all_consensus" > srst2_AMR.log
```

### 2. Uncertainty assessment of allele calls

This step aims to determine whether a dubious allele call (denoted by a question mark "?", such as "FloR\_1212*?" and "SulI\_1616?") can be treated as a reliable allele call. For example, given a merged gene profile generated by the command line:


```
python srst2/scripts/srst2.py --prev_output *_demo__genes__ARGannot_r2__results.txt --output Demo
```

| Sample  | FloR\_Phe    | SulI\_Sul    |
|---------|-------------|-------------|
| strain1 | FloR\_1212\*? | SulI\_1616\*? |
| strain2 | FloR\_1212\*  | SulI\_1616?  |

where three dubious allele calls are present. Supposing we can determine that SulI\_1616\*? and SulI\_1616? are reliable while FloR\_1212\*? is unreliable, then the gene profile becomes

| Sample  | FloR\_Phe    | SulI\_Sul    |
|---------|-------------|-------------|
| strain1 | - | SulI\_1616\* |
| strain2 | FloR\_1212\*  | SulI\_1616  |

We can use scripts under the sub-directory reliability\_assessment of PAMmaker to perform this uncertainty assessment. The procedure is comprised of two steps.

```
python /vlsci/SG0006/shared/wan/scripts/PAMmaker/reliability_assessment/collate_topAllele_scores.py --allele_calls ./Gene_calls/*_aEPEC__genes__ARGannot_r2__results.txt --allele_scores ./Scores/*_aEPEC__*.ARGannot_r2.scores --prefix mergedScores
```



```
Rscript /vlsci/SG0006/shared/wan/scripts/PAMmaker/reliability_assessment/assessAlleleCallUncertainty.R --profiles aEPEC__compiledResults.txt --scores mergedScores__gene.scores --output aEPEC_srst2__reliableCalls
```